src/utils/transaction.ts

```diff
@@ -0,0 +1,130 @@
+import { JsonRpcTx } from "../deps.ts";
+import {
+  AccessListEIP2930Transaction,
+  bytesToHex,
+  FeeMarketEIP1559Transaction,
+  intToHex,
+  isAccessListEIP2930Tx,
+  isFeeMarketEIP1559TxData,
+  isLegacyTx,
+  LegacyTransaction,
+  TransactionFactory,
+  TypedTransaction,
+  TypedTxData,
+} from "../deps.ts";
+
+/**
+ * @param tx Typed transaction to be signed.
+ * @param r Signature r value.
+ * @param s Signature s value.
+ * @param v Signature v value. In case of EIP155ReplayProtection, must include the chain ID.
+ * @returns Passed transaction with the signature added.
+ * @throws Error if the transaction is a BlobEIP4844Tx or if v param is < 35 for a
+ *         LegacyTx.
+ */
+export function addSignature(
+  tx: TypedTransaction,
+  r: bigint,
+  s: bigint,
+  v: bigint,
+): TypedTransaction {
+  const TypedTxData = ((): TypedTxData => {
+    if (isLegacyTx(tx)) {
+      if (v < 35) {
+        throw new Error(`Invalid v value: ${v}`);
+      }
+      return LegacyTransaction.fromTxData({
+        nonce: tx.nonce,
+        gasPrice: tx.gasPrice,
+        gasLimit: tx.gasLimit,
+        to: tx.to,
+        value: tx.value,
+        data: tx.data,
+        v,
+        r,
+        s,
+      });
+    } else if (isAccessListEIP2930Tx(tx)) {
+      return AccessListEIP2930Transaction.fromTxData({
+        chainId: tx.chainId,
+        nonce: tx.nonce,
+        gasPrice: tx.gasPrice,
+        gasLimit: tx.gasLimit,
+        to: tx.to,
+        value: tx.value,
+        data: tx.data,
+        accessList: tx.accessList,
+        v,
+        r,
+        s,
+      });
+    } else if (isFeeMarketEIP1559TxData(tx)) {
+      return FeeMarketEIP1559Transaction.fromTxData({
+        chainId: tx.chainId,
+        nonce: tx.nonce,
+        maxPriorityFeePerGas: tx.maxPriorityFeePerGas,
+        maxFeePerGas: tx.maxFeePerGas,
+        gasLimit: tx.gasLimit,
+        to: tx.to,
+        value: tx.value,
+        data: tx.data,
+        accessList: tx.accessList,
+        v,
+        r,
+        s,
+      });
+    } else {
+      throw new Error(`Invalid transaction type: ${tx}`);
+    }
+  })();
+
+  return TransactionFactory.fromTxData(
+    TypedTxData,
+  );
+}
+
+type Option<T> = T | null;
```

Comment:
I'd be inclined to not use this but the idiomatic typescript way of T | null, it's easier

Reply:
ok!